crates/runtime/src/dataconnector/github.rs

```diff
@@ -0,0 +1,324 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use crate::component::dataset::Dataset;
+use async_trait::async_trait;
+use data_components::graphql::{client::GraphQLClient, provider::GraphQLTableProvider};
+use datafusion::datasource::TableProvider;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::{any::Any, error::Error, future::Future, pin::Pin, sync::Arc};
+use url::Url;
+
+use super::{
+    graphql::default_spice_client, DataConnector, DataConnectorFactory, ParameterSpec, Parameters,
+};
+
+pub struct Github {
+    params: Parameters,
+}
+
+pub trait GithubTableArgs: Send + Sync {
+    /// Converts the arguments for a Github table into a tuple of:
+    ///   1. The GraphQL query string
+    ///   2. The JSON pointer to the data in the response
+    ///   3. The depth to unnest the data
+    fn get_graphql_values(&self) -> (String, String, usize);
+}
+
+// TODO: implement PR filters from https://docs.github.com/en/graphql/reference/objects#repository `Arguments for pullRequests`.
+pub struct PullRequestTableArgs {
+    pub owner: String,
+    pub repo: String,
+}
+
+impl GithubTableArgs for PullRequestTableArgs {
+    fn get_graphql_values(&self) -> (String, String, usize) {
+        let query = format!(
+            r#"
+            {{
+                repository(owner: "{owner}", name: "{name}") {{
+                    pullRequests(first: 100) {{
+                        pageInfo {{
+                            hasNextPage
+                            endCursor
+                        }}
+                        nodes {{
+                            title
+                            url
+                            body
+                            state
+                            createdAt
+                            mergedAt
+                            closedAt
+                            number
+                            num_of_reviews: reviews {{totalCount}}
+                            ref: headRef {{ id }}
+                            
+                            author {{
+                                login
+                            }}
+                            additions
+                            deletions
+                            changedFiles
+                            comments(first: 100) {{num_of_comments: totalCount}}
+                            commits(first: 100) {{num_of_commits: totalCount, hashes: nodes{{ id }} }}
```

Comment:
Wondering if we can flatten hashes from a list of structs to a list of Utf8. This should simplify working with them. (if there is only id/utf8 field)

Reply:
No, i tried that. In graphQL there is no great way to do this. We may need to consider being able to export views ontop of data connectors. Alternatively, we might be able to do this in GraphQL specifically since its using a `MemTable`.